sdb/commands/spl/internal/kmem_helpers.py

```diff
@@ -116,20 +121,79 @@ def entry_size(cache: drgn.Object) -> int:
 
 
 def active_memory(cache: drgn.Object) -> int:
-    assert cache.type_.type_name() == 'spl_kmem_cache_t *'
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
     return obj_alloc(cache) * entry_size(cache)
 
 
 def total_memory(cache: drgn.Object) -> int:
-    assert cache.type_.type_name() == 'spl_kmem_cache_t *'
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
     if backed_by_linux_cache(cache):
         return slub.total_memory(cache.skc_linux_cache)
     return slab_size(cache) * nr_slabs(cache)
 
 
 def util(cache: drgn.Object) -> int:
-    assert cache.type_.type_name() == 'spl_kmem_cache_t *'
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
     total_mem = total_memory(cache)
     if total_mem == 0:
         return 0
     return int((active_memory(cache) / total_mem) * 100)
+
+
+def sko_from_obj(cache: drgn.Object, obj: drgn.Object) -> drgn.Object:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    return sdb.create_object(
+        'spl_kmem_obj_t *',
+        obj.value_() + p2.p2roundup(object_size(cache), cache_obj_align))
+
+
+def spl_aligned_obj_size(cache: drgn.Object) -> int:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    spl_obj_type_size = sdb.type_canonicalize_size('spl_kmem_obj_t')
+    return p2.p2roundup(object_size(cache), cache_obj_align) + p2.p2roundup(
+        spl_obj_type_size, cache_obj_align)
+
+
+def spl_aligned_slab_size(cache: drgn.Object) -> int:
+    assert sdb.type_canonical_name(cache.type_) == 'struct spl_kmem_cache *'
+    cache_obj_align = cache.skc_obj_align.value_()
+    spl_slab_type_size = sdb.type_canonicalize_size('spl_kmem_slab_t')
+    return p2.p2roundup(spl_slab_type_size, cache_obj_align)
+
+
+def for_each_onslab_object_in_slab(slab: drgn.Object) -> Iterable[drgn.Object]:
+    assert sdb.type_canonical_name(slab.type_) == 'struct spl_kmem_slab *'
+    cache = slab.sks_cache
+    sks_size = spl_aligned_slab_size(cache)
+    spl_obj_size = spl_aligned_obj_size(cache)
+
+    for i in range(slab.sks_objs.value_()):
+        obj = sdb.create_object('void *',
+                                slab.value_() + sks_size + (i * spl_obj_size))
+        #
+        # If the sko_list of the object is empty, it means that
+        # this object is not part of the slab's internal free list
+        # and therefore it is allocated.
+        #
+        sko = sko_from_obj(cache, obj)
+        assert sko.sko_magic.value_() == 0x20202020  # SKO_MAGIC
+        if linked_lists.is_list_empty(sko.sko_list):
```

Comment:
My understanding is that `sko_list` is not a list, but rather a link on a list.  And `is_list_empty` is not telling us if this "list" is empty, but rather if the link is on any list.  Most of this is the fault of the linux kernel's choice of terminology, so I'm not sure how much we can do about this besides adding a comment (which is sort-of covered above but I think we could be more explicit).

Reply:
Added a more explicit note about this in the existing comment.